tlscommon.c

```diff
@@ -626,13 +628,199 @@ static int certattr_matchip(GENERAL_NAME *gn, struct certattrmatch *match){
 }
 
+static char* concatenate(char* s1, char* s2) {
+    if(s1 == NULL) {
+        return s2;
+    } 
+    if(s2 == NULL) {
+        return s1;
+    }
+    if(s1 == NULL && s2 == NULL) {
+        return NULL;
+    }
+    char* result = (char*) calloc(strlen(s1) + strlen(s2), sizeof(char)); 
+    if(result == NULL) {
+        return NULL;
+    }
+    memcpy(result, s1, strlen(s1));
+    memcpy(result + strlen(s1), s2, strlen(s2));
+    return result;
+}
+
+static char* append(int n, ...) {
+   char* val = NULL;
+   va_list ap;
+   int i;
+
+   va_start(ap, n);
+   char* result = NULL;
+   for(i = 0; i < n; i++) {
+      val = va_arg(ap, char*);
+      result = concatenate(result, val);
+   }
+   va_end(ap);
+   return result;
+}
+
+static char* extractFirstPortion(char* dns, char c) {
+    char* result = NULL;
+    char *firstPos = strchr(dns, c);
+    if(firstPos == NULL) {
+        return result;
+    }
+    int lenOfFirstPortion = strlen(dns) - strlen(firstPos);
+    result = (char*) calloc(lenOfFirstPortion, sizeof(char));
+    if(result == NULL) {
+        return NULL;
+    }
+    memcpy(result, dns, lenOfFirstPortion);
+    return result;
+}
+
+static int regexMatcher(char* pattern, char* str) {
+    regex_t    regex;
+    char buffer[100];
+    
+    int reti = regcomp(&regex, pattern, REG_EXTENDED);
+    if (reti) {
+        regerror(reti, &regex, buffer, 100);
+        debug(DBG_ERR, "regcomp failed with '%s'\n", buffer);
+        return 0;
+    }        
+    reti = regexec(&regex, str, 0, NULL, 0);
+    if(reti) {             
+        return 0;                                            
+    }
+    return 1;
+}
+
+static char* calculateLenOfLabel(int len) {
+    char str[3];
+    sprintf(str, "%d", len);
+    debug(DBG_DBG, "str : %s", str);
+    char* result = append(3, "{0,", str, "}");
+    return result;
+}
+
+// compareWithModifiedHostname generates a regular expression based on the first label present
+// in certificate DNS. Matches the regular expression with the first label of hostname and
+// returns result - 1 - matched, 0 - unmatched
+// Refer https://datatracker.ietf.org/doc/html/rfc6125#section-6.4
+// buzz* => ^buzz[a-zA-Z0-9-]{0,59}$
+// *buzz => ^[a-z0-9A-Z][a-z0-9A-Z-]{0,59}buzz$
+// b*uzz => ^b[a-zA-Z0-9-]{0,59}uzz$
+// 63 - Maximum length of a label 
+// Wiki: A domain name consists of one or more labels, each of which is formed from the set of ASCII letters, 
+// digits, and hyphens (a-z, A-Z, 0–9, -), but not starting or ending with a hyphen. 
+// The labels are case-insensitive
+// Internationlized Domain Names: www.äöü.com converted to www.xn--4ca0bs.com. xn--4ca0bs is considered
+// a valid DN and let through. diseñolatinoamericano.com => xn--diseolatinoamericano-66b.com will also be
+// let through. So IDNs are passed in Punycode encoded and xn-- says everything that follows is unicode encoded.
+static int compareWithModifiedHostname(char* certDNS, char* hostname) {
```

Comment:
Karthika, there is a much safer memory pattern to use here.   The reason for the pattern below is that it keeps all memory deallocations in one place for better readability, and better maintainability.  This is 'C' version of exception handling (sort of).  Consider the following pattern...

```
char* firstPorCertDNS = NULL;
char* firstPorHostName = NULL;
char* modifiedHostName = NULL;
int retcode = ERROR;

....
do stuff
....
if (I detected and error) {
  goto cleanup;
}

... do more stuff ...
   retcode = 1;

   return retcode;

cleanup:
  if (firstPorCertDNS != NULL) {
    free(firstPorCertDNS);
    firstPorCertDNS = NULL;
  }

  if (firstPorHostName != NULL) {
    free(firstPorHostName);
    firstPorHostName = NULL;
  }
 
  if (modifiedHostName != NULL) {
    free(modifiedHostName);
    modifiedHostName = NULL;
  }
  
  return retcode;
}